add index fail for "too many sst files are ingesting"

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

add index
flashback to the time before add index 
during flashback inject tidb failure and tikv-pd lantency

### 2. What did you expect to see? (Required)
add index successfully
### 3. What did you see instead (Required)
add index fail 
[2023/05/23 23:11:34.790 +00:00] [WARN] [index.go:1020] ["[ddl] run add index job failed, convert job to rollback"] [job="ID:291, Type:add index, State:running, SchemaState:write reorganization, SchemaID:89, TableID:109, RowCount:10000000, ArgLen:6, start time: 2023-05-23 22:29:45.107 +0000 UTC, Err:[Lightning:KV:ServerIsBusy]too many sst files are ingesting, ErrCount:1, SnapshotVersion:441681441038270473, UniqueWarnings:0"] [error="[ddl:8214]Cancelled DDL job"]
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
7.1.0

Normally, TiDB should retry even if "too many sst" is reported. However, there is a panic during retry:

```
[2023/05/23 23:11:34.790 +00:00] [ERROR] [misc.go:116] ["panic in the recoverable goroutine"] [label=ddl] [funcInfo=onCreateIndex] [r="\"invalid memory address or nil pointer dereference\""] [stack="github.com/pingcap/tidb/util.Recover\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/misc.go:120\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:884\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:260\nruntime.sigpanic\n\t/usr/local/go/src/runtime/signal_unix.go:837\ngithub.com/pingcap/tidb/br/pkg/lightning/backend.(*OpenedEngine).Flush\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/backend.go:270\ngithub.com/pingcap/tidb/ddl/ingest.(*engineInfo).Flush\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/ingest/engine.go:83\ngithub.com/pingcap/tidb/ddl/ingest.(*litBackendCtx).Flush\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/ingest/backend.go:163\ngithub.com/pingcap/tidb/ddl/ingest.(*CheckpointManager).Sync\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/ingest/checkpoint.go:219\ngithub.com/pingcap/tidb/ddl.(*ingestBackfillScheduler).close\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/backfilling_scheduler.go:334\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:884\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:260\nruntime.sigpanic\n\t/usr/local/go/src/runtime/signal_unix.go:837\ngithub.com/pingcap/tidb/br/pkg/lightning/backend.(*OpenedEngine).LocalWriter\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/backend.go:275\ngithub.com/pingcap/tidb/ddl/ingest.(*engineInfo).newWriterContext\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/ingest/engine.go:198\ngithub.com/pingcap/tidb/ddl/ingest.(*engineInfo).CreateWriter\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/ingest/engine.go:173\ngithub.com/pingcap/tidb/ddl.newAddIndexIngestWorker\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/index.go:1616\ngithub.com/pingcap/tidb/ddl.(*ingestBackfillScheduler).createWorker\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/backfilling_scheduler.go:404\ngithub.com/pingcap/tidb/resourcemanager/pool/workerpool.(*WorkerPool[...]).runAWorker\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/resourcemanager/pool/workerpool/workerpool.go:107\ngithub.com/pingcap/tidb/resourcemanager/pool/workerpool.NewWorkerPool[...]\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/resourcemanager/pool/workerpool/workerpool.go:91\ngithub.com/pingcap/tidb/ddl.(*ingestBackfillScheduler).setupWorkers\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/backfilling_scheduler.go:311\ngithub.com/pingcap/tidb/ddl.(*ddlCtx).writePhysicalTableRecord\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/backfilling.go:710\ngithub.com/pingcap/tidb/ddl.(*worker).addPhysicalTableIndex\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/index.go:1788\ngithub.com/pingcap/tidb/ddl.(*worker).addTableIndex\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/index.go:1836\ngithub.com/pingcap/tidb/ddl.runReorgJobAndHandleErr.func2\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/index.go:1004\ngithub.com/pingcap/tidb/ddl.(*worker).runReorgJob.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/ddl/reorg.go:253"]
```

This is because the engine is closed but it doesn't get cleaned from backend context:

```
[2023/05/23 22:30:10.854 +00:00] [INFO] [engine.go:120] ["[ddl-ingest] flush all writer and get closed engine"] ["job ID"=291] ["index ID"=3]
```

As a result, the engine is not opened in the next retry.

https://github.com/pingcap/tidb/blob/ad0957fbc1a16d6ffdbfb75410f76c7ea508d62c/ddl/ingest/engine_mgr.go#LL46C3-L46C3

--- a/ddl/index.go
+++ b/ddl/index.go
@@ -779,6 +779,8 @@ func cleanupSortPath(currentJobID int64) error {
 		}
 		if _, ok := ingest.LitBackCtxMgr.Load(jobID); ok {
 			// The job is still running, skip it.
+			logutil.BgLogger().Warn("[ddl-ingest] the job is still running, skip removing it",
+				zap.Int64("running job ID", jobID))
 			continue
 		}
 		// Remove all the temp data of the previous done jobs.

--- a/ddl/index.go
+++ b/ddl/index.go
@@ -1847,6 +1847,10 @@ func (w *worker) addTableIndex(t table.Table, reorgInfo *reorgInfo) error {
 			if err != nil {
 				return errors.Trace(err)
 			}
+			// Every time we finish a partition, we update the progress of the job.
+			if rc := w.getReorgCtx(reorgInfo.Job.ID); rc != nil {
+				reorgInfo.Job.SetRowCount(rc.getRowCount())
+			}
 		}
 	} else {
 		//nolint:forcetypeassert

--- a/ddl/index.go
+++ b/ddl/index.go
@@ -777,8 +777,11 @@ func cleanupSortPath(currentJobID int64) error {
 			logutil.BgLogger().Warn("[ddl-ingest] cannot cleanup sort path", zap.Error(err))
 			continue
 		}
-		// For now, there is only one task using ingest at the same time,
-		// so we can remove all the temp data of the previous jobs.
+		if _, ok := ingest.LitBackCtxMgr.Load(jobID); ok {
+			// The job is still running, skip it.
+			continue
+		}
+		// Remove all the temp data of the previous done jobs.
 		if jobID < currentJobID {
 			logutil.BgLogger().Info("[ddl-ingest] remove stale temp index data",
 				zap.Int64("jobID", jobID), zap.Int64("currentJobID", currentJobID))

--- a/ddl/index.go
+++ b/ddl/index.go
@@ -921,6 +921,8 @@ func runIngestReorgJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job,
 	done, ver, err = runReorgJobAndHandleErr(w, d, t, job, tbl, indexInfo, false)
 	if err != nil {
 		if !errorIsRetryable(err, job) {
+			logutil.BgLogger().Warn("[ddl] run reorg job failed, convert job to rollback",
+				zap.String("job", job.String()), zap.Error(err))
 			ver, err = convertAddIdxJob2RollbackJob(d, t, job, tbl.Meta(), indexInfo, err)
 			ingest.LitBackCtxMgr.Unregister(job.ID)
 		}

--- a/ddl/ingest/BUILD.bazel
+++ b/ddl/ingest/BUILD.bazel
@@ -43,6 +43,7 @@ go_library(
         "//util/size",
         "@com_github_google_uuid//:uuid",
         "@com_github_pingcap_errors//:errors",
+        "@com_github_pingcap_failpoint//:failpoint",
         "@io_etcd_go_etcd_client_v3//:client",
         "@io_etcd_go_etcd_client_v3//concurrency",
         "@org_uber_go_atomic//:atomic",

--- a/ddl/ingest/backend.go
+++ b/ddl/ingest/backend.go
@@ -19,6 +19,7 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/br/pkg/lightning/backend/encode"
 	"github.com/pingcap/tidb/br/pkg/lightning/backend/local"
 	lightning "github.com/pingcap/tidb/br/pkg/lightning/config"
@@ -121,6 +122,10 @@ func (bc *litBackendCtx) FinishImport(indexID int64, unique bool, tbl table.Tabl
 	if err != nil {
 		return err
 	}
+
+	failpoint.Inject("mockFinishImportErr", func() {
+		failpoint.Return(fmt.Errorf("mock finish import error"))
+	})
 
 	// Check remote duplicate value for the index.
 	if unique {

--- a/ddl/ingest/engine_mgr.go
+++ b/ddl/ingest/engine_mgr.go
@@ -36,7 +36,7 @@ func (bc *litBackendCtx) Register(jobID, indexID int64, schemaName, tableName st
 
 	var info string
 	en, exist := bc.Load(indexID)
-	if !exist {
+	if !exist || en.writerCount == 0 {
 		engineCacheSize := int64(bc.cfg.TikvImporter.EngineMemCacheSize)
 		ok := bc.MemRoot.CheckConsume(StructSizeEngineInfo + engineCacheSize)
 		if !ok {

--- a/ddl/ingest/engine_mgr.go
+++ b/ddl/ingest/engine_mgr.go
@@ -36,7 +36,7 @@ func (bc *litBackendCtx) Register(jobID, indexID int64, schemaName, tableName st
 
 	var info string
 	en, exist := bc.Load(indexID)
-	if !exist || en.writerCount == 0 {
+	if !exist || en.openedEngine == nil {
 		engineCacheSize := int64(bc.cfg.TikvImporter.EngineMemCacheSize)
 		ok := bc.MemRoot.CheckConsume(StructSizeEngineInfo + engineCacheSize)
 		if !ok {

--- a/tests/realtikvtest/addindextest/integration_test.go
+++ b/tests/realtikvtest/addindextest/integration_test.go
@@ -369,3 +369,25 @@ func TestAddIndexSplitTableRanges(t *testing.T) {
 	tk.MustExec("admin check table t;")
 	ddl.SetBackfillTaskChanSizeForTest(1024)
 }
+
+func TestAddIndexFinishImportError(t *testing.T) {
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("drop database if exists addindexlit;")
+	tk.MustExec("create database addindexlit;")
+	tk.MustExec("use addindexlit;")
+	tk.MustExec(`set global tidb_ddl_enable_fast_reorg=on;`)
+
+	tk.MustExec("create table t (a int primary key, b int);")
+	for i := 0; i < 4; i++ {
+		tk.MustExec(fmt.Sprintf("insert into t values (%d, %d);", i*10000, i*10000))
+	}
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/ingest/mockFinishImportErr", "1*return"))
+	tk.MustExec("alter table t add index idx(a);")
+	require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/ddl/ingest/mockFinishImportErr"))
+	tk.MustExec("admin check table t;")
+	rows := tk.MustQuery("admin show ddl jobs 1;").Rows()
+	//nolint: forcetypeassert
+	jobTp := rows[0][3].(string)
+	require.True(t, strings.Contains(jobTp, "ingest"), jobTp)
+}

